### PR TITLE
Use sets to speed up queryRenderedFeatures

### PR DESCRIPTION
**Pull request by tomhicks**
## What this does

Fixes https://github.com/maplibre/maplibre-gl-js/issues/4770

Uses a Set instead of an Array for internals that traverse the "filter layer IDs" that a user passes when querying rendered features. This speeds up `queryRenderedFeatures` significantly when we have a large number of layers, as `Array.prototype.indexOf` becomes slow with large numbers of elements.

We convert to a Set as soon as possible, which means that some other semi-public APIs (e.g. `Tile::queryRenderedFeatures`) have changed to require a Set instead of an Array.

## Launch Checklist

- [X] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
- [X] Briefly describe the changes in this PR.
- [X] Link to related issues.
- [X] Include before/after visuals or gifs if this PR includes visual changes.
- [X] Write tests for all new functionality.
- [X] Document any changes to public APIs.
- [X] Post benchmark scores.
- [X] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

File: src/style/style.ts [hunk 1]
```diff
@@ -1433,6 +1434,12 @@ export class Style extends Evented {
         // LayerSpecification is serialized StyleLayer, and this casting is safe.
         const serializedLayers = this._serializedAllLayers() as {[_: string]: StyleLayer};
 
+        const layersAsSet = params.layers instanceof Set ? params.layers : Array.isArray(params.layers) ? new Set(params.layers) : undefined;
```
Comment (HarelM) [inline src/style/style.ts:1437]: Wouldn't this code create performance issues when calling `queryRenderedFeatures` "fast"? For example when using mouse hover?

File: src/source/query_features.ts [hunk 2]
```diff
@@ -7,6 +7,8 @@ import type {FilterSpecification} from '@maplibre/maplibre-gl-style-spec';
 import type {MapGeoJSONFeature} from '../util/vectortile_to_geojson';
 import type Point from '@mapbox/point-geometry';
 import {mat4} from 'gl-matrix';
+import {FeatureIndex} from '../data/feature_index';
```
Comment (HarelM) [inline src/source/query_features.ts:10]: `import type ...`?

File: src/source/query_features.ts [hunk 3]
```diff
@@ -102,7 +108,7 @@ export function queryRenderedFeatures(
     for (const tileIn of tilesIn) {
         renderedFeatureLayers.push({
             wrappedTileID: tileIn.tileID.wrapped().key,
-            queryResults: tileIn.tile.queryRenderedFeatures(
+            queryResults: (tileIn.tile as Tile).queryRenderedFeatures(
```
Comment (HarelM) [inline src/source/query_features.ts:111]: Why was this added?

File: src/source/query_features.ts [hunk 4]
```diff
@@ -16,7 +18,7 @@ export type QueryRenderedFeaturesOptions = {
      * An array of [style layer IDs](https://maplibre.org/maplibre-style-spec/#layer-id) for the query to inspect.
      * Only features within these layers will be returned. If this parameter is undefined, all layers will be checked.
      */
-    layers?: Array<string>;
+    layers?: Array<string> | Set<string>;
```
Comment (HarelM) [inline src/source/query_features.ts:21]: Please update the comment above this field.

**Comment by HarelM** [comment]
Can you add some test for this?

File: src/source/query_features.ts (hunk 3 above)
Comment (tomhicks) [reply src/source/query_features.ts:111]: Because without it the queryRenderedFeatures call is undiscoverable by TypeScript. I found that in a few places I ended up with some uncovered TS (types resolving to `any`) and these assertions were the minimal change to get some amount of type safety back.

File: src/style/style.ts (hunk 1 above)
Comment (tomhicks) [reply src/style/style.ts:1437]: This will always be faster than the resulting calls into the `queryRenderedFeatures` helper functions, as they have a tendency to iterate over themselves. This will be O(n) and makes the rest of the call stack O(n) (often a loop of length n with an O(1) lookup inside) as opposed to O(n^2).

Also, the user can escape this conversion if they really need to by passing their own Set.

Comment (HarelM) [reply src/style/style.ts:1437]: Fair enough.

File: src/source/query_features.ts (hunk 3 above)
Comment (HarelM) [reply src/source/query_features.ts:111]: Is it possible to improve the typing that lead to this problem instead of adding `as ...`?
I know it's a bigger change, I know, but probably better...?

File: src/source/query_features.ts (hunk 2 above)
Comment (tomhicks) [reply src/source/query_features.ts:10]: It might be a good idea to enable that check in the tsconfig to avoid having to manually inspect things like this.

File: src/style/style.ts (hunk 1 above)
Comment (tomhicks) [reply src/style/style.ts:1437]: Here's an example showing this performance gain, even if we convert to a set every time. In MapLibre the gains would be even more than this because there are I think at least two times where the arrays are iterated in a nested fashion.

https://jsperf.app/xekege

File: src/source/query_features.ts (hunk 2 above)
Comment (HarelM) [reply src/source/query_features.ts:10]: I'm not familiar with it, feel free to push this forward.

File: src/data/feature_index.ts [hunk 5]
```diff
@@ -196,7 +196,7 @@ export class FeatureIndex {
         ) => boolean | number) {
 
         const layerIDs = this.bucketLayerIDs[bucketIndex];
-        if (filterLayerIDs && !arraysIntersect(filterLayerIDs, layerIDs))
+        if (filterLayerIDs && !layerIDs.filter(i => filterLayerIDs.has(i)))
```
Comment (neodescis) [inline src/data/feature_index.ts:199]: Did you mean to check the length of the filtered `layerIDs` array here? The `.filter()` function will always return something truthy.

Comment (tomhicks) [reply src/data/feature_index.ts:199]: er yes this is supposed to be `some` or `every` - good spot. I'll also try to add a breaking test.

File: src/source/source_cache.ts [hunk 6]
```diff
@@ -948,9 +948,15 @@ export class SourceCache extends Evented {
      * @param pointQueryGeometry - coordinates of the corners of bounding rectangle
      * @returns result items have `{tile, minX, maxX, minY, maxY}`, where min/max bounding values are the given bounds transformed in into the coordinate space of this tile.
      */
-    tilesIn(pointQueryGeometry: Array<Point>, maxPitchScaleFactor: number, has3DLayer: boolean): any[] {
-
-        const tileResults = [];
+    tilesIn(pointQueryGeometry: Array<Point>, maxPitchScaleFactor: number, has3DLayer: boolean) {
+        type TileResult = {
```
Comment (HarelM) [inline src/source/source_cache.ts:952]: Can this be defined outside the method? Preferably in the upper part of the file.

File: src/data/feature_index.ts [hunk 7]
```diff
@@ -113,7 +113,7 @@ export class FeatureIndex {
     ): {[_: string]: Array<{featureIndex: number; feature: GeoJSONFeature}>} {
         this.loadVTLayers();
 
-        const params = args.params || {} as { filter: any; layers: string[]; availableImages: string[] },
+        const params = args.params || {} as QueryParameters['params'],
```
Comment (HarelM) [inline src/data/feature_index.ts:116]: Use a type for this, it will help both here and in the above definition...

**Comment by HarelM** [comment]
Besides my nit-picking, this looks good.
THANKS!

**Comment by tomhicks** [comment]
I've tried enabling `noImplicitAny` to highlight the places where `any`s are getting into the code and it's a lot of places, so I won't be including that in this PR.

I will look at opening another one or two with [`noImplicitAny`](https://www.typescriptlang.org/tsconfig/#noImplicitAny) and [`verbatimModuleSyntax`](https://www.typescriptlang.org/tsconfig/#verbatimModuleSyntax) which should reduce nits and improve type safety.

**Comment by HarelM** [comment]
No explicit any is a great idea, but as you said, should be out of scope of this PR.

File: src/source/source_cache.ts [hunk 8]
```diff
@@ -948,9 +948,15 @@ export class SourceCache extends Evented {
...
+            tile: Tile;
+            tileID: OverscaledTileID;
+            queryGeometry: Array<Point>;
+            cameraQueryGeometry: Array<Point>;
+            scale: number;
+        };
+        const tileResults: Array<TileResult> = [];
```
Comment (neodescis) [inline src/source/source_cache.ts:959]:
```suggestion
        const tileResults: TileResult[] = [];
```
I think we should prefer [] over Array<>, at least for simple types

File: src/data/feature_index.ts [hunk 9]
```diff
@@ -196,7 +196,7 @@ export class FeatureIndex {
         ) => boolean | number) {
 
         const layerIDs = this.bucketLayerIDs[bucketIndex];
-        if (filterLayerIDs && !arraysIntersect(filterLayerIDs, layerIDs))
+        if (filterLayerIDs && !layerIDs.some(id => filterLayerIDs.has(id)))
```
Comment (ibesora) [inline src/data/feature_index.ts:199]: Awesome work @tomhicks !

Have you considered the alternative approach of converting `bucketLayerIDs[bucketIndex]`to a `Set` instead?
It wouldn't require any change to any of the public signatures and get similar results by reversing the arrays in your condition to something like 
```typescript
if (filterLayerIDs && !filterLayerIDs.some(id => layerIDs.has(id)))
```

Comment (ibesora) [reply src/data/feature_index.ts:199]: Oh, although that would get similar results, your following improvement would not be possible 🤦 
Nothing to see here.

Comment (tomhicks) [reply src/data/feature_index.ts:199]: It may be beneficial to convert them both to sets at some point and use [Set.prototype.isDisjointFrom](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/isDisjointFrom) but that's very new so would need a shim for older browsers.

Also depends on whether there are any genuine array semantics necessary, such as mapping, filtering, etc.

**Comment by tomhicks** [comment]
One thing I'm not quite sure of here is how to take this to completion by including some kind of regression test for performance.

The unit tests for `style.queryRenderedFeatures` use a very heavily stubbed map which wouldn't show the performance benefit. I notice there are benchmarks in `test/bench` but I'm not sure how to set up a very heavy map that would show this benefit, again.

I wonder if someone could help me do this? @ibesora as you're familiar with the "bike lanes" layer, this perf issue pops up if you've got say 100 of the bike lanes layers and you're querying around northern Europe. The density of features seems to be a factor aside from just number of layers.

File: src/data/feature_index.ts (hunk 9 above)
Comment (tomhicks) [reply src/data/feature_index.ts:199]: FYI here are some performance comparisons for set intersection implementations: https://jsperf.app/modede

It might be worth considering this when there are other cases of indexOf usage where the size of the arrays can get very large. In smaller cases, the overhead of generating and accessing a Set will be slower than indexOf, so it really depends whether the dimensions are likely to get large or can get large.

**Comment by ibesora** [comment]
@tomhicks I just pushed a commit to your branch that adds a benchmark. 
Instead of going all the way through from `style.queryRenderedFeatures` it just tests `featureIndex.loadMatchingFeature` which is where the performance changes happened.

Here it is a comparison without and with your improvements. This is testing a tile with 100 layers filtering by 100 non-existing layer ids, which is the worst case scenario we had before.
Screenshot 2024-10-04 at 10 12 17
tomsImprovements

Your implementation is basically an order of magnitude faster than what we currently have 💪

**Comment by tomhicks** [comment]
thanks for adding that @ibesora - good to know it shows up in benchmarks, too.

I think this is pretty much good to go now?
- all the requested changes are addressed
- it performs better
- the important public APIs aren't changed
- it has tests that cover the new Array-or-Set APIs

File: src/source/source_cache.ts [hunk 10]
```diff
@@ -1100,3 +1099,11 @@ function compareTileId(a: OverscaledTileID, b: OverscaledTileID): number {
 function isRasterType(type) {
     return type === 'raster' || type === 'image' || type === 'video';
 }
+
+type TileResult = {
```
Comment (HarelM) [inline src/source/source_cache.ts:1103]: Can this be moved to the beginning of the file please?

**Review by HarelM — APPROVED** [review]

File: src/source/source_cache.ts (hunk 10 above)
Comment (tomhicks) [reply src/source/source_cache.ts]: uff, just pushed. thought i'd pushed this already

**Comment by HarelM** [comment]
Can you add a changelog item?

**Comment by tomhicks** [comment]
@HarelM changelog item added. I think that's all the things from the checklist.

**Review by HarelM — APPROVED** [review]

**Comment by HarelM** [comment]
THANKS!
